### PR TITLE
Add the k nearest p-median module and the tutorial example for capacitated p-median

### PR DESCRIPTION
**Pull request by rongboxu**
Hi all! This pull request originates from the feature branch and includes all the commits from PR #387 . It can be identified as the outcome of the GSOC 2023 project.

### CODE REVIEW

File: .ci/310.yaml [hunk 1]
```diff
@@ -15,6 +15,7 @@ dependencies:
   - shapely
   - spaghetti
   - tqdm
+  - pointpats=2.3
```
Comment (jGaboardi) [inline .ci/310.yaml:18]: Let's remove the pin here (and all other `.ci/` environments ***except*** `.ci/38-MIN.yaml`):

```suggestion
  - pointpats
```

File: .ci/311-DEV.yaml [hunk 2]
```diff
@@ -14,6 +14,7 @@ dependencies:
   - scipy
   - spaghetti
   - tqdm
+  - pointpats=2.3
```
Comment (jGaboardi) [inline .ci/311-DEV.yaml:17]:
```suggestion
  - pointpats
```

File: .ci/311.yaml [hunk 3]
```diff
@@ -14,6 +14,7 @@ dependencies:
   - scipy
   - spaghetti
   - tqdm
+  - pointpats=2.3
```
Comment (jGaboardi) [inline .ci/311.yaml:17]:
```suggestion
  - pointpats
```

File: .ci/39.yaml [hunk 4]
```diff
@@ -14,6 +14,7 @@ dependencies:
   - scipy
   - shapely
   - spaghetti
+  - pointpats=2.3
```
Comment (jGaboardi) [inline .ci/39.yaml:17]:
```suggestion
  - pointpats
```

**Comment by jGaboardi** [comment]
I am not sure it is a `scikit-learn` compatibility issue, because current `pointpats` testing passes with [`scikit-learn==1.3.0`](https://github.com/pysal/pointpats/actions/runs/5920587962/job/16052052378#step:3:1174). Did tests pass locally for you?

**Comment by rongboxu** [comment]
> I am not sure it is a `scikit-learn` compatibility issue, because current `pointpats` testing passes with [`scikit-learn==1.3.0`](https://github.com/pysal/pointpats/actions/runs/5920587962/job/16052052378#step:3:1174). Did tests pass locally for you?

My local python is 3.8 version, so the test can pass.

**Comment by rongboxu** [comment]
I am not very sure, but maybe this can be the reason of the failing of `build_best_tree`, see [this](https://github.com/pysal/pointpats/pull/107).
The relevant code in the Github is, `if metric in kdtree_valid_metrics:`, see [here.](https://github.com/pysal/pointpats/blob/main/pointpats/geometry.py#L320C1-L321C1)
While, in the test detail, it is `if metric in KDTree.valid_metrics:`, see [here.](https://github.com/pysal/spopt/actions/runs/5939258287/job/16105296901?pr=397#step:6:475)

**Comment by gegen07** [comment]
I think [`pointpats`](https://github.com/pysal/pointpats/releases/tag/v2.3.0) was not released after this bug fix by Martin, then it throws this error. I think we should release it first to merge this pull request.

**Comment by jGaboardi** [comment]
@gegen07 @ljwolf , with the new [`pointpats` release](https://github.com/pysal/pointpats/releases) all tests are passing except the [known AZP failure](https://github.com/pysal/spopt/issues/386). I think we are good to merge this. Any dissenting  opinions?

**Comment by gegen07** [comment]
I think it is good to go.

**Comment by jGaboardi** [comment]
Congrats, @rongboxu!

**Comment by rongboxu** [comment]
> Congrats, @rongboxu!

Thank you very much for merging it!

**Pull request merged**